### PR TITLE
Fix stange behavior of DownloadPullDiffOrPatch in incorect index

### LINKED ISSUE #17105: API returns strange things when getting `api/v1/repos/{owner}/{repo}/pulls/{index}.{any extensions except diff or patch}`
### Gitea Version

1.16.0+dev-267-g5ac857f4d

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

Built from source code

### Database

_No response_

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Description

If I enter an URL like https://codeberg.org/api/v1/repos/gitnex/Gitnex/pulls/975.test, it returns a strange result, the PR with the number 139. It should return a 404 error.

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by pricly-yellow**
Close #17105

### CODE REVIEW

**Review by a1012112796 — APPROVED** [review]

**Comment by a1012112796** [comment]
suggest update test code:
```DIFF
diff --git a/models/pull_test.go b/models/pull_test.go
index 2b7ef2f66..173977aaf 100644
--- a/models/pull_test.go
+++ b/models/pull_test.go
@@ -134,6 +134,10 @@ func TestGetPullRequestByIndex(t *testing.T) {
 	_, err = GetPullRequestByIndex(9223372036854775807, 9223372036854775807)
 	assert.Error(t, err)
 	assert.True(t, IsErrPullRequestNotExist(err))
+
+	_, err = GetPullRequestByIndex(1, 0)
+	assert.Error(t, err)
+	assert.True(t, IsErrPullRequestNotExist(err))
 }
 
 func TestGetPullRequestByID(t *testing.T) {

```

**Comment by pricly-yellow** [comment]
Good idea, push it now

File: models/pull.go [hunk 1]
```diff
@@ -522,6 +522,9 @@ func GetLatestPullRequestByHeadInfo(repoID int64, branch string) (*PullRequest,
 
 // GetPullRequestByIndex returns a pull request by the given index
 func GetPullRequestByIndex(repoID, index int64) (*PullRequest, error) {
+	if index < 1 {
```
Comment (delvh) [inline models/pull.go:525]: For more experienced maintainers: Is `0` a valid PR ID or not?
If yes, this `if` should be updated.

File: models/pull_test.go [hunk 2]
```diff
@@ -134,6 +134,10 @@ func TestGetPullRequestByIndex(t *testing.T) {
 	_, err = GetPullRequestByIndex(9223372036854775807, 9223372036854775807)
 	assert.Error(t, err)
 	assert.True(t, IsErrPullRequestNotExist(err))
+
+	_, err = GetPullRequestByIndex(1, 0)
```
Comment (delvh) [inline models/pull_test.go:138]: Depending on the outcome above, this line might need to be refactored.

**Review by delvh — COMMENTED** [review]
Apart from this question: LGTM.

File: models/pull.go (hunk 1 above)
Comment (6543) [reply models/pull.go:525]: index starts with 1

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
the (xorm) Get will ignore fields with "default init values" and match with set one
```go
var value int64
fmt.Println(value)
```
That's why Get is mostly used with the ".ID()" function witch we can not use since `index` is a normal field for xorm (we do use it ID'like for each repo)

So either check with if if it is 0 and return 404 or add a explisit Where("index=?", pr.index). In this case an if is more elegant since 0 is not an valide index

@delvh (:point_up:)

**Comment by 6543** [comment]
@pricly-yellow thanks for the pull, can you cherry-pick squash-commit  to release branch 1.15 and make a pull?

if you need help just tell

**Comment by 6543** [comment]
-> #17227

**Pull request merged**